File: problem-types/five_knights_checkmate.js
function autoscroll(x, y) {
	block_nav()
	let [x_add, y_add] = [100, 100]
	let [y_tmp, x_tmp] = ['no', 'no']
	let dur = {'left': -1, 'right': 1, 'up': -1, 'down': 1, 'no': 0}
	$(scroll_p).css({'left': x, 'top': y - -y_add})
	if (!in_access_zone(scroll_p))
		y_tmp = 'down'

	$(scroll_p).css({'left': x, 'top': y - y_add})
	if (!in_access_zone(scroll_p))
		y_tmp = 'up'

	$(scroll_p).css({'top': y, 'left': x - -x_add})
	if (!in_access_zone(scroll_p))
		x_tmp = 'right'

	$(scroll_p).css({'top': y, 'left': x - x_add})
	if (!in_access_zone(scroll_p))
		x_tmp = 'left'

	scrollBy(dur[x_tmp] * x_add / 2, dur[y_tmp] * y_add / 2)

}

function in_window(obj) {
	let rect = $(obj)[0].getBoundingClientRect()
	let [hor_add, vert_add] = [rect.width / 2, rect.height / 2]
	if (rect.left + hor_add < 0 || rect.right - hor_add > $(window).width())
		return false
	if (rect.top + vert_add < 0 || rect.bottom - vert_add > $(window).height())
		return false
	return true
}

function is_intersect(obj1, obj2) {
	let [rect1, rect2] = [$(obj1)[0].getBoundingClientRect(), $(obj2)[0].getBoundingClientRect()]
	if (rect1.left > rect2.right || rect2.left > rect1.right) 
		return false
	if (rect1.bottom < rect2.top || rect2.bottom < rect1.top)
		return false
	return true
}

function in_access_zone(obj) {
	let nav = $('nav.user_nav')
	if (is_intersect(obj, nav))
		return false
	if (!in_window(obj))
		return false
	return true
}

function getSVGCoordinates(event) {
	if (event.touches) {
		let coordinatePoint = new DOMPoint(event.touches[0].clientX, event.touches[0].clientY)
		let svg = document.querySelector('svg')
		return coordinatePoint.matrixTransform(svg.getScreenCTM().inverse())
	}
	else {
		let coordinatePoint = new DOMPoint(event.clientX, event.clientY)
		let svg = document.querySelector('svg')
		return coordinatePoint.matrixTransform(svg.getScreenCTM().inverse())
	}
}

function occupied(ind, only_horse = false) {
	let [x, y] = [$(`rect:eq(${ind})`).attr('x'), $(`rect:eq(${ind})`).attr('y')]
	let tmp = 0
	if (only_horse)
		$('image.horse.active').each(function(index){if ($(this).attr('x') == x && $(this).attr('y') == y) tmp = 1})
	else
		$('image').each(function(index){if ($(this).attr('x') == x && $(this).attr('y') == y) tmp = 1})
	return tmp == 1
}
function move(e) {
	let obj = $('.targeted')
	let svg = $('svg')[0].getBoundingClientRect()
	if (!in_access_zone(obj)) {
		end_move(in_access=false)
	}
	else {
		autoscroll(e.clientX, e.clientY)
		$(obj).attr({'x': getSVGCoordinates(e).x - side / 2,
					'y': getSVGCoordinates(e).y - side / 2})
	}
}

function create_new_horse() {
	let obj = $('.horse')
	let new_horse = $(document.createElementNS("http://www.w3.org/2000/svg", 'image'))
	    .addClass('horse active')
	    .attr({
	        x: def_x,
	        y: def_y,
	        width: $(obj).attr('width'),
	        height: $(obj).attr('height'),
	        href: $(obj).attr('href')
	    });
	    $('svg').append(new_horse)
}

function back_to_drag() {
	$(document).off('mousemove touchmove')
	$('.targeted').remove()
	remain += 1
}

function start_move(e) {
	let obj = $(e.currentTarget) 
	if (!$(obj).hasClass('choiced')) {
		remain -= 1
	}
	$('svg').append(obj)
	let svg = $('svg')[0].getBoundingClientRect()
	$(obj).attr({'x': getSVGCoordinates(e).x - side / 2,
				'y': getSVGCoordinates(e).y - side / 2})
	$(obj).addClass('targeted')
	$(document).on('mousemove touchmove', move)
	update_status()
}

function end_move(in_access=true) {
	let obj = $('.targeted')
	let [best_ind, min_dist] = [0, Math.sqrt(2) * side]
	$('rect').each(function(index){
		let dist = Math.hypot(($(obj).attr('x') - $(this).attr('x')), ($(obj).attr('y') - $(this).attr('y')))
		if (min_dist >= dist)
			[best_ind, min_dist] = [index, dist]
	})
	if (best_ind == '' || occupied(best_ind) || !in_access) {
		back_to_drag()
	}
	else {
		$(document).off('mousemove touchmove')
		$(obj).attr('x', $(`rect:eq(${best_ind})`).attr('x')).attr('y', $(`rect:eq(${best_ind})`).attr('y'))
		$(obj).addClass('choiced')
		$(obj).removeClass('targeted')
	}
	if (remain > 0 && $('.active').not('.choiced').length == 0) create_new_horse()
	update_status()
}

function update_status() {
	$('.horse.active').off('mousedown touchstart')
	$('.horse.active').off('mouseup touchend')
	$('.horse.active').on('mousedown touchstart', start_move)
	$('.horse.active').on('mouseup touchend', end_move)
	$('text.amount')[0].innerHTML = remain
}

var side = $('.horse.active').attr('width')
var [def_x, def_y] = [$('.horse.passive.drag').attr('x'), $('.horse.passive.drag').attr('y')]
var remain = $('text.amount').html()

$('.horse.active').on('mousedown touchstart', start_move)
$('.horse.active').on('mouseup touchend', end_move)
$('.reload').on('click touchstart', function(){
	$('.horse.active').remove()
	create_new_horse()
	remain = 4
	update_status()
})
$('.submit_button').on('click touchstart', function(e){
	if (e.touches)
		e.preventDefault()
	let ans = ''
	$('rect').each(function(index){
		if (occupied(index, only_horse=true)) ans += "1,"
		else ans += "'',"
	})
	$('input[name="answer"]').val(ans)
})

var scroll_p = $(document.createElement('div')).addClass('scroll_div')
$('body').append(scroll_p)

File: problem-types/five_knights_checkmate.py
import sys

def entry_form(data, kwargs):
	height = 8 # число клеточек в столбце
	width = 8 # число клеточек в строке
	line_width = 2 # ширина линий 
	ind = line_width // 2 # отступ
	side = 70 # длина стороны квадрата с границами
	inner_side = side - line_width # длина стороны квадрата без границ 
	board_width = line_width + side * width # длина доски
	board_height = line_width + side * height # ширина доски
	pad = 6 # расстояние между доской и зоной перетаскивания
	plot_width = pad + board_width + inner_side 
	plot_height = board_height
	yield '<input name="answer" type="hidden" />'
	yield '<div class="plot_area">'
	yield f'<svg version="1.1" width="{plot_width}" height="{plot_height}" overflow="visible" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">'
	for y in range(0, height + 1):
		yield f'<line class="grid_line" x1="{ind}" y1 = "{ind + y * side}" x2="{ind + width * side}" y2="{ind + y * side}" stroke-width="{line_width}"/>'
	for x in range(0, width + 1):
		yield f'<line class="grid_line" x1="{ind + x * side}" y1 = "{ind}" x2="{ind + x * side}" y2="{ind + height * side}" stroke-width="{line_width}"/>'
	for y in range(0, height):
		for x in range(0, width):
			if (x + y) % 2 == 1:
				yield f'<rect class="orange" x="{x * side + line_width}" y="{y * side + line_width}" width="{inner_side}" height="{inner_side}" />'
			else:
				yield f'<rect class="white" x="{x * side + line_width}" y="{y * side + line_width}" width="{inner_side}" height="{inner_side}" />'
	yield f'<image class="king passive" x="{(data["king_pos"][1] - 1) * side + line_width}" y="{(data["king_pos"][0] - 1) * side + line_width}" width="{inner_side}" height="{inner_side}" href="/static/chess/king_b.png" />'
	yield f'<image class="horse passive" x="{(data["horse_pos"][1] - 1) * side + line_width}" y="{(data["horse_pos"][0] - 1) * side + line_width}" width="{inner_side}" height="{inner_side}" href="/static/chess/horse_w.png" />'
	
	yield f'<image class="horse passive drag" x="{pad + board_width}" y="{line_width}" width="{inner_side}" height="{inner_side}" href="/static/chess/horse_w.png" />'
	yield f'<image class="horse active" x="{pad + board_width}" y="{line_width}" width="{inner_side}" height="{inner_side}" href="/static/chess/horse_w.png" />'
	yield f'<text style="font-size: {inner_side * 0.2}px" textLength="{inner_side * 0.8}" x="{pad + board_width}" y="{inner_side * 1.2}">Осталось</text>'
	yield f'<text class="amount" style="font-size: {inner_side * 0.2}px" textLength="{inner_side * 0.1}" x="{pad + board_width + inner_side * 0.9}" y="{inner_side * 1.2}"> 4</text>'
	yield f'<image class="reload" x = "{pad + board_width}" y="{plot_height - inner_side}" height="{inner_side}" width="{inner_side}" href="/static/reload.png" />'
	yield '</svg>'
	yield '</div>'

def validate(data, answer):
	ans = list(answer[:-1].split(','))
	ans = ['' if i == "''" else i for i in ans]
	for var in data['correct']:
		tmp = 1
		for i in range(8):
			for j in range(8):
				if str(ans[i * 8 + j]) != str(var[i][j]):
					tmp = 0
		if tmp == 1:
			return True
	return False